src/sync/async_open_task.cpp

```diff
@@ -0,0 +1,74 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2019 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "impl/realm_coordinator.hpp"
+#include "sync/async_open_task.hpp"
+#include "sync/sync_manager.hpp"
+#include "sync/sync_session.hpp"
+
+namespace realm {
+
+AsyncOpenTask::AsyncOpenTask(std::string realm_path):
+ m_coordinator(_impl::RealmCoordinator::get_coordinator(realm_path)),
+ m_session(SyncManager::shared().get_existing_session(realm_path))
+{
+}
+
+void AsyncOpenTask::start(std::function<void(std::shared_ptr<Realm>, std::exception_ptr)> callback) {
+    m_session->wait_for_download_completion([callback, this](std::error_code ec) {
```

Comment:
This callback needs to capture a weak_ptr to the task and verify that it still exists. Currently calling `cancel()` and then destroying the AsyncOpenTask will result in a use-after-free when the download actually finishes.